taskcat/_lambda_build.py

```diff
@@ -1,52 +1,67 @@
 from __future__ import print_function
 
 import logging
-import os
-from multiprocessing import Pool
-from shutil import make_archive
+from pathlib import Path
+from uuid import UUID, uuid5
 
-from taskcat._common_utils import make_dir
+import docker
+
+from ._config import Config
 
 LOG = logging.getLogger(__name__)
 
 
 class LambdaBuild:
-    """Zips contents of lambda source files.
-
-    """
-
-    zip_file_name = "lambda"
-
-    def __init__(self, source_path, output_path="../packages", threads=4):
-
-        cur_dir = os.path.abspath(os.curdir)
-        try:
-            self.source_path = os.path.abspath(source_path)
-            os.chdir(self.source_path)
-            self.output_path = os.path.abspath(output_path)
-
-            dirs = [i for i in os.listdir("./") if os.path.isdir(i)]
-            pool = Pool(threads)
-            pool.map(self._make_zip, dirs)
-            pool.close()
-            pool.join()
-        finally:
-            os.chdir(cur_dir)
-
-    def _make_zip(self, name):
-        try:
-            LOG.info("Zipping lambda function %s" % name)
-            output_path = "%s/%s" % (self.output_path, name)
-            make_dir(output_path)
-            os.chdir(name)
-            make_archive(output_path + "/" + self.zip_file_name, "zip", "./")
-        finally:
-            os.chdir(self.source_path)
+    NULL_UUID = UUID("{00000000-0000-0000-0000-000000000000}")
+
+    def __init__(self, config: Config, build_submodules=True):
+        self._docker = docker.from_env()
+        self._config = config
+        self._build_lambdas(config.lambda_source_path, config.lambda_zip_path)
+        if build_submodules:
+            rel_source = config.lambda_source_path.relative_to(config.project_root)
+            rel_zip = config.lambda_zip_path.relative_to(config.project_root)
+            self._recurse(config.project_root, rel_source, rel_zip)
+
+    def _recurse(self, base_path, rel_source, rel_zip):
+        submodules_path = Path(base_path) / "submodules"
+        if submodules_path.is_dir():
+            for submodule in submodules_path.iterdir():
+                source_path = submodule / rel_source
+                output_path = submodule / rel_zip
+                if source_path.is_dir():
+                    self._build_lambdas(source_path, output_path)
+                self._recurse(submodule, rel_source, rel_zip)
+
+    def _build_lambdas(self, parent_path: Path, output_path):
+        if parent_path.is_dir:
+            for path in parent_path.iterdir():
+                if (path / "Dockerfile").is_file():
+                    tag = f"taskcat-build-{uuid5(self.NULL_UUID, str(path)).hex}"
+                    LOG.info(
+                        f"Packaging lambda source from {path} using docker image {tag}"
+                    )
+                    self._docker_build(path, tag)
+                    self._docker_extract(tag, output_path / path.stem)
+
+    @staticmethod
+    def _clean_build_log(line):
```

Comment:
I had a crazy thought about a nested dict.get that cascades into the next value as a default, but on second thought it smells too much like magic and is less clear.

That said, this may be a good use case for the ternary operator + dict.get. Or not.

There’s something about it, but can’t really think of anything clearer. Just leaving this because I already typed it in case it jogs something. But ignore otherwise.

Reply:
how about

```
    def _clean_build_log(line):
        line = line.get("stream", line.get("aux", line))
        return str(line).strip()
```

shorter, but probably less clear. Leaving as is for now

Reply:
yes, that's kinda what I had thought. but as you said it may be less clear without having to spend some time parsing the line. leaving it as is seems best